packages/devtools_app/integration_test/test_infra/test_app_driver.dart

```diff
@@ -0,0 +1,244 @@
+    final String wsUriString = debugPort['params']['wsUri'];
+    _vmServiceWsUri = Uri.parse(wsUriString);
+
+    // Map to WS URI.
+    _vmServiceWsUri =
+        convertToWebSocketUrl(serviceProtocolUrl: _vmServiceWsUri);
+
+    await started;
+  }
+
+  Future<int> killGracefully() async {
+    _debugPrint('Sending SIGTERM to $runProcessId..');
+    Process.killPid(runProcessId);
+
+    final killFuture =
+        runProcess!.exitCode.timeout(_quitTimeout, onTimeout: _killForcefully);
+    unawaited(_killAndShutdown(killFuture));
+    return killFuture;
+  }
+
+  Future<int> _killForcefully() async {
+    _debugPrint('Sending SIGKILL to $runProcessId..');
+    Process.killPid(runProcessId, ProcessSignal.sigkill);
+
+    final killFuture = runProcess!.exitCode;
+    unawaited(_killAndShutdown(killFuture));
+    return killFuture;
+  }
+
+  Future<void> _killAndShutdown(Future<int> killFuture) async {
+    unawaited(
+      killFuture.then((_) async {
+        await cancelAllStreamSubscriptions();
+      }),
+    );
+  }
+
+  Future<Map<String, dynamic>> waitFor({
+    String? event,
+    int? id,
+    Duration? timeout,
+    bool ignoreAppStopEvent = false,
+  }) async {
+    final Completer<Map<String, dynamic>> response =
+        Completer<Map<String, dynamic>>();
+    late StreamSubscription<String> sub;
+    sub = stdoutController.stream.listen((String line) async {
+      final dynamic json = _parseFlutterResponse(line);
+      if (json == null) {
+        return;
+      } else if ((event != null && json['event'] == event) ||
+          (id != null && json['id'] == id)) {
+        await sub.cancel();
+        response.complete(json);
+      } else if (!ignoreAppStopEvent && json['event'] == 'app.stop') {
+        await sub.cancel();
+        final StringBuffer error = StringBuffer();
+        error.write('Received app.stop event while waiting for ');
+        error.write(
+          '${event != null ? '$event event' : 'response to request $id.'}.\n\n',
+        );
+        if (json['params'] != null && json['params']['error'] != null) {
+          error.write('${json['params']['error']}\n\n');
+        }
+        if (json['params'] != null && json['params']['trace'] != null) {
+          error.write('${json['params']['trace']}\n\n');
+        }
+        response.completeError(error.toString());
+      }
+    });
+
+    return _timeoutWithMessages<Map<String, dynamic>>(
+      () => response.future,
+      timeout: timeout,
+      message: event != null
+          ? 'Did not receive expected $event event.'
+          : 'Did not receive response to request "$id".',
+    ).whenComplete(() => sub.cancel());
+  }
+
+  Future<T> _timeoutWithMessages<T>(
+    Future<T> Function() f, {
+    Duration? timeout,
+    String? message,
+  }) {
+    // Capture output to a buffer so if we don't get the response we want we can show
+    // the output that did arrive in the timeout error.
+    final StringBuffer messages = StringBuffer();
```

Comment:
Nit: consider removing LHS types for declarations where it's clear what the type should be.